xwiki-platform-core/xwiki-platform-attachment/xwiki-platform-attachment-validation/xwiki-platform-attachment-validation-default/src/main/java/org/xwiki/attachment/validation/internal/DefaultAttachmentValidator.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.attachment.validation.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.regex.Pattern;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import javax.servlet.http.Part;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.xwiki.attachment.validation.AttachmentValidationConfiguration;
+import org.xwiki.attachment.validation.AttachmentValidationException;
+import org.xwiki.attachment.validation.AttachmentValidator;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.tika.internal.TikaUtils;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.web.UploadAction;
+
+import static com.xpn.xwiki.plugin.fileupload.FileUploadPlugin.UPLOAD_DEFAULT_MAXSIZE;
+import static com.xpn.xwiki.plugin.fileupload.FileUploadPlugin.UPLOAD_MAXSIZE_PARAMETER;
+import static javax.servlet.http.HttpServletResponse.SC_REQUEST_ENTITY_TOO_LARGE;
+import static javax.servlet.http.HttpServletResponse.SC_UNSUPPORTED_MEDIA_TYPE;
+import static org.apache.commons.lang.exception.ExceptionUtils.getRootCauseMessage;
+
+/**
+ * Default implementation of {@link AttachmentValidator}. Check for the file size and mimetype of a given file. The
+ * configuration values are retrieved using {@link AttachmentValidationConfiguration}.
+ *
+ * @version $Id$
+ * @since 14.10RC1
+ */
+@Component
+@Singleton
+public class DefaultAttachmentValidator implements AttachmentValidator
+{
+    @Inject
+    @Named("readonly")
```

Comment:
Using a readonly context for perf reason?

Reply:
Mainly because currently there is no reason to do anything else than reading the context in this class.

Reply:
Ok, wondering about the perf impact of using one instead of the other: pretty sure that we could use `readonly` context in many more components. @tmortagne an idea about that?

Reply:
The point of the "readonly" XWikiContext provider is not really performances. As documented on ReadOnlyXWikiContextProvider the only difference is that, if there is no XWikiContext in the current ExecutionContext, you won't get a generated one, so you need to be prepared for the eventuality that it might return null (which I don't think is the case here).